File: Python/Algoritmo Genético.py
import random
import time


# Función para calcular la distancia entre dos puntos
def distancia(p1, p2):
    return ((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2) ** 0.5


# Función para calcular el largo de un recorrido
def largo_recorrido(ruta, puntos):
    return sum(distancia(puntos[ruta[i]], puntos[ruta[i + 1]]) for i in range(len(ruta) - 1))


# Operador de cruce (crossover)
def cruce(padre1, padre2):
    start, end = sorted(random.sample(range(len(padre1)), 2))
    hijo = [-1] * len(padre1)
    hijo[start:end + 1] = padre1[start:end + 1]

    idx = 0
    for i in range(len(padre2)):
        if padre2[i] not in hijo:
            while hijo[idx] != -1:
                idx += 1
            hijo[idx] = padre2[i]

    return hijo


# Algoritmo Genético
def algoritmo_genetico(puntos, n_generaciones=500, n_poblacion=100, tasa_mutacion=0.1):
    poblacion = [random.sample(range(len(puntos)), len(puntos)) for _ in range(n_poblacion)]

    for _ in range(n_generaciones):
        poblacion = sorted(poblacion, key=lambda ruta: largo_recorrido(ruta, puntos))
        nueva_poblacion = poblacion[:2]  # Seleccionamos los dos mejores

        while len(nueva_poblacion) < n_poblacion:
            padre1, padre2 = random.sample(poblacion[:n_poblacion // 2], 2)
            hijo = cruce(padre1, padre2)
            if random.random() < tasa_mutacion:
                i, j = random.sample(range(len(hijo)), 2)
                hijo[i], hijo[j] = hijo[j], hijo[i]
            nueva_poblacion.append(hijo)

        poblacion = nueva_poblacion

    mejor_ruta = min(poblacion, key=lambda ruta: largo_recorrido(ruta, puntos))
    return mejor_ruta, largo_recorrido(mejor_ruta, puntos)


# Ejemplo de uso
if __name__ == "__main__":
    puntos = [(random.randint(0, 100), random.randint(0, 100)) for _ in range(10)]

    start_time = time.time()
    mejor_ruta, mejor_largo = algoritmo_genetico(puntos)
    end_time = time.time()

    print("Mejor ruta:", mejor_ruta)
    print("Distancia total:", mejor_largo)
    print("Tiempo de ejecución:", end_time - start_time, "segundos")

File: c++/Algoritmo Genético.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <algorithm>  // Para std::find, std::iota, std::shuffle, std::sort
#include <random>     // Para random_device, mt19937, uniform_int_distribution
#include <limits>     // Para numeric_limits

using namespace std;

// Función de distancia
double distancia(pair<int, int> p1, pair<int, int> p2) {
    return sqrt(pow(p1.first - p2.first, 2) + pow(p1.second - p2.second, 2));
}

// Función para calcular el largo de un recorrido
double largo_recorrido(const vector<int>& ruta, const vector<pair<int, int>>& puntos) {
    double total = 0;
    for (size_t i = 0; i < ruta.size() - 1; ++i) {
        total += distancia(puntos[ruta[i]], puntos[ruta[i + 1]]);
    }
    return total;
}

// Operador de cruce (crossover)
vector<int> cruce(const vector<int>& padre1, const vector<int>& padre2) {
    size_t start = rand() % padre1.size();
    size_t end = rand() % padre1.size();
    if (start > end) swap(start, end);

    vector<int> hijo(padre1.size(), -1);
    for (size_t i = start; i <= end; ++i) {
        hijo[i] = padre1[i];
    }

    size_t idx = 0;
    for (size_t i = 0; i < padre2.size(); ++i) {
        if (find(hijo.begin(), hijo.end(), padre2[i]) == hijo.end()) {
            while (hijo[idx] != -1) idx++;
            hijo[idx] = padre2[i];
        }
    }

    return hijo;
}

// Algoritmo Genético
pair<vector<int>, double> algoritmo_genetico(vector<pair<int, int>>& puntos, int n_generaciones = 500, int n_poblacion = 100, double tasa_mutacion = 0.1) {
    vector<vector<int>> poblacion;
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> distrib(0, puntos.size() - 1);

    // Inicialización de la población
    for (int i = 0; i < n_poblacion; ++i) {
        vector<int> individuo(puntos.size());
        iota(individuo.begin(), individuo.end(), 0);
        shuffle(individuo.begin(), individuo.end(), gen);
        poblacion.push_back(individuo);
    }

    vector<int> mejor_ruta;
    double mejor_largo = numeric_limits<double>::infinity();

    for (int g = 0; g < n_generaciones; ++g) {
        // Ordenar la población por el largo del recorrido
        sort(poblacion.begin(), poblacion.end(), [&puntos](const vector<int>& a, const vector<int>& b) {
            return largo_recorrido(a, puntos) < largo_recorrido(b, puntos);
        });

        vector<vector<int>> nueva_poblacion(poblacion.begin(), poblacion.begin() + 2);

        while (nueva_poblacion.size() < n_poblacion) {
            int i = rand() % (n_poblacion / 2);
            int j = rand() % (n_poblacion / 2);
            vector<int> hijo = cruce(poblacion[i], poblacion[j]);
            if ((rand() % 100) < tasa_mutacion * 100) {
                swap(hijo[rand() % puntos.size()], hijo[rand() % puntos.size()]);
            }
            nueva_poblacion.push_back(hijo);
        }

        poblacion = nueva_poblacion;

        // Evaluar la mejor ruta en la población
        for (const auto& ruta : poblacion) {
            double largo = largo_recorrido(ruta, puntos);
            if (largo < mejor_largo) {
                mejor_largo = largo;
                mejor_ruta = ruta;
            }
        }
    }

    return make_pair(mejor_ruta, mejor_largo);
}

int main() {
    srand(time(0));

    vector<pair<int, int>> puntos;
    for (int i = 0; i < 10; ++i) {
        puntos.push_back({rand() % 100, rand() % 100});
    }

    clock_t start_time = clock();

    // Uso tradicional de 'first' y 'second' para obtener los resultados del pair
    pair<vector<int>, double> resultado = algoritmo_genetico(puntos);
    vector<int> mejor_ruta = resultado.first;
    double mejor_largo = resultado.second;

    clock_t end_time = clock();

    cout << "Mejor ruta:";
    for (int i : mejor_ruta) {
        cout << " " << i;
    }
    cout << "\nDistancia total: " << mejor_largo << "\n";
    cout << "Tiempo de ejecución: " << double(end_time - start_time) / CLOCKS_PER_SEC << " segundos\n";
}
